agent/logging.go

```diff
@@ -185,26 +192,68 @@ func (a *Agent) uploadToS3(ctx context.Context, tc *taskContext) error {
 		return errors.Wrap(err, "error creating pail")
 	}
 
-	return a.uploadLogFiles(ctx, tc, bucket)
+	return a.uploadLogDir(ctx, tc, bucket, "")
 }
 
-func (a *Agent) uploadLogFiles(ctx context.Context, tc *taskContext, bucket pail.Bucket) error {
+func (a *Agent) uploadLogDir(ctx context.Context, tc *taskContext, bucket pail.Bucket, name string) error {
 	if tc.taskConfig == nil || tc.taskConfig.Task == nil {
 		return nil
 	}
 	catcher := grip.NewBasicCatcher()
-	catcher.Add(a.uploadSingleFile(ctx, bucket, agentLogFileName, tc.taskConfig.Task.Id, tc.taskConfig.Task.Execution))
-	catcher.Add(a.uploadSingleFile(ctx, bucket, systemLogFileName, tc.taskConfig.Task.Id, tc.taskConfig.Task.Execution))
-	catcher.Add(a.uploadSingleFile(ctx, bucket, taskLogFileName, tc.taskConfig.Task.Id, tc.taskConfig.Task.Execution))
+	path := filepath.Join(a.opts.WorkingDirectory, taskLogDirectory)
+	if name != "" {
+		path = filepath.Join(path, name)
+	}
+	dir, err := ioutil.ReadDir(path)
+	if err != nil {
+		catcher.Add(errors.Wrap(err, "error reading log directory"))
+		return catcher.Resolve()
+	}
+	for _, f := range dir {
+		if f.IsDir() {
+			catcher.Add(a.uploadLogDir(ctx, tc, bucket, f.Name()))
+		} else {
+			catcher.Add(a.uploadSingleFile(ctx, tc, bucket, f.Name(), tc.taskConfig.Task.Id, tc.taskConfig.Task.Execution, name))
+		}
+	}
 
 	return catcher.Resolve()
 }
 
-func (a *Agent) uploadSingleFile(ctx context.Context, bucket pail.Bucket, file string, taskID string, execution int) error {
-	localPath := filepath.Join(a.opts.WorkingDirectory, taskLogDirectory, file)
+func (a *Agent) uploadSingleFile(ctx context.Context, tc *taskContext, bucket pail.Bucket, file string, taskID string, execution int, command string) error {
+	localDir := filepath.Join(a.opts.WorkingDirectory, taskLogDirectory)
+	remotePath := fmt.Sprintf("logs/%s/%s", taskID, strconv.Itoa(execution))
+	if command != "" {
+		localDir = filepath.Join(localDir, command)
+		remotePath = fmt.Sprintf("%s/%s", remotePath, command)
+	}
+	localPath := filepath.Join(localDir, file)
 	_, err := os.Stat(localPath)
 	if os.IsNotExist(err) {
 		return nil
 	}
-	return bucket.Upload(ctx, filepath.Join("logs", taskID, strconv.Itoa(execution), file), localPath)
+	err = bucket.Upload(ctx, fmt.Sprintf("%s/%s", remotePath, file), localPath)
```

Comment:
why not filepath join this?

Reply:
This is a URL even though logically it's a filepath. It won't work on Windows